ops.h

```diff
@@ -551,47 +551,66 @@ static HWY_NOINLINE HWY_MAYBE_UNUSED void RopeAndMulBy(const float mul,
 }
 
 static HWY_NOINLINE HWY_MAYBE_UNUSED void AddFrom(
-    const float* HWY_RESTRICT other, float* HWY_RESTRICT x, size_t size) {
-  for (size_t i = 0; i < size; ++i) {
-    x[i] += other[i];
-  }
+    const float* HWY_RESTRICT other, float* HWY_RESTRICT x, const size_t size) {
+  namespace hn = hwy::HWY_NAMESPACE;
+  using D = hn::ScalableTag<float>;
+  const D d;
+
+  hn::Transform1(d, x, size, other,
+                 [](const auto d, const auto x, const auto other)
+                     HWY_ATTR { return hn::Add(x, other); });
 }
 
 static HWY_NOINLINE void MulBy(const float* HWY_RESTRICT other,
-                               float* HWY_RESTRICT x, size_t size,
-                               size_t max_pos) {
+                               float* HWY_RESTRICT x, const size_t size,
+                               const size_t max_pos) {
   HWY_DASSERT(max_pos <= size);
-  for (size_t i = 0; i < max_pos; ++i) {
-    x[i] *= other[i];
-  }
+  namespace hn = hwy::HWY_NAMESPACE;
+  using D = hn::ScalableTag<float>;
+  const D d;
+
+  hn::Transform1(d, x, max_pos, other,
+                 [](const auto d, const auto x, const auto other)
+                     HWY_ATTR { return hn::Mul(x, other); });
 }
 
 static HWY_INLINE HWY_MAYBE_UNUSED void MulBy(const float* HWY_RESTRICT other,
                                               float* HWY_RESTRICT x,
-                                              size_t size) {
+                                              const size_t size) {
   return MulBy(other, x, size, size);
 }
 
-static HWY_NOINLINE void MulByConst(float c, float* HWY_RESTRICT x, size_t size,
-                                    size_t max_pos) {
+static HWY_NOINLINE void MulByConst(const float c, float* HWY_RESTRICT x,
+                                    const size_t size, const size_t max_pos) {
   HWY_DASSERT(max_pos <= size);
-  for (size_t i = 0; i < max_pos; ++i) {
-    x[i] *= c;
-  }
+  namespace hn = hwy::HWY_NAMESPACE;
+  using D = hn::ScalableTag<float>;
+  const D d;
+  const auto constant = hn::Set(d, c);
```

Comment:
Unfortunately this breaks on SVE, current compilers do not allow capturing references to vectors. How about passing in the `float c` and calling Set inside the lambda? Would be good to verify via gcc.godbolt.org that the codegen is still reasonable i.e., the compiler is hoisting Set out of the loop.

If not, we can consider adding TransformWithConstant or a similar function that passes the scalar constant in without a lambda capture, and passes the result of Set to the lambda. If we go that route, to ensure the code still works with older Highway versions, we could add a Highway version check and `#if` too old, then fall back to the previous plain C++ code.

Reply:
Thank you @jan-wassenberg , I had seen your comment [here](https://github.com/google/gemma.cpp/blob/8fb44ed6dd123f63dca95c20c561e8ca1de511d7/ops.h#L613)
But I have tried this code in [godbolt](https://gcc.godbolt.org/) with these flags `-O2 -march=armv8-a+sve` with Arm64 gcc-10.2 it was ok,
```

HWY_ATTR void MulByConst(const float c, float* HWY_RESTRICT x,
                                    const size_t size, const size_t max_pos) {
  HWY_DASSERT(max_pos <= size);
  namespace hn = hwy::HWY_NAMESPACE;
  using D = hn::ScalableTag<float>;
  const D d;
  const auto constant = hn::Set(d, c);
  hn::Transform(d, x, max_pos,
                [&constant](const auto d, const auto x)
                    HWY_ATTR { return hn::Mul(x, constant); });
}
``` 
Maybe I miss something.

Anyway, I will fix them :)

Reply:
Thanks for checking back. I see you are right, this indeed works on Clang and GCC. The key difference is the capture by ref - by value indeed fails to compile. We can keep your code as-is then :)

Would you also like to remove the comment you mentioned, and use copy-by-ref + Foreach there?

Reply:
Thanks , Yes sure, I will do it